core/src/main/scala/org/apache/spark/deploy/security/HadoopDelegationTokenManager.scala

```diff
@@ -17,76 +17,175 @@
  */
 private[spark] class HadoopDelegationTokenManager(
-    sparkConf: SparkConf,
-    hadoopConf: Configuration,
-    fileSystems: Configuration => Set[FileSystem])
-  extends Logging {
+    protected val sparkConf: SparkConf,
+    protected val hadoopConf: Configuration) extends Logging {
 
   private val deprecatedProviderEnabledConfigs = List(
     "spark.yarn.security.tokens.%s.enabled",
     "spark.yarn.security.credentials.%s.enabled")
   private val providerEnabledConfig = "spark.security.credentials.%s.enabled"
 
-  // Maintain all the registered delegation token providers
-  private val delegationTokenProviders = getDelegationTokenProviders
+  private val principal = sparkConf.get(PRINCIPAL).orNull
+  private val keytab = sparkConf.get(KEYTAB).orNull
+
+  if (principal != null) {
+    require(keytab != null, "Kerberos principal specified without a keytab.")
+    require(new File(keytab).isFile(), s"Cannot find keytab at $keytab.")
+  }
+
+  private val delegationTokenProviders = loadProviders()
   logDebug("Using the following builtin delegation token providers: " +
     s"${delegationTokenProviders.keys.mkString(", ")}.")
 
-  /** Construct a [[HadoopDelegationTokenManager]] for the default Hadoop filesystem */
-  def this(sparkConf: SparkConf, hadoopConf: Configuration) = {
-    this(
-      sparkConf,
-      hadoopConf,
-      hadoopConf => Set(FileSystem.get(hadoopConf).getHomeDirectory.getFileSystem(hadoopConf)))
+  private var renewalExecutor: ScheduledExecutorService = _
+  private val driverRef = new AtomicReference[RpcEndpointRef]()
+
+  protected def setDriverRef(ref: RpcEndpointRef): Unit = {
+    driverRef.set(ref)
   }
 
-  private def getDelegationTokenProviders: Map[String, HadoopDelegationTokenProvider] = {
-    val providers = Seq(new HadoopFSDelegationTokenProvider(fileSystems)) ++
-      safeCreateProvider(new HiveDelegationTokenProvider) ++
-      safeCreateProvider(new HBaseDelegationTokenProvider)
+  /**
+   * Start the token renewer. Upon start, if a principal has been configured, the renewer will:
+   *
+   * - log in the configured principal, and set up a task to keep that user's ticket renewed
+   * - obtain delegation tokens from all available providers
+   * - schedule a periodic task to update the tokens when needed.
+   *
+   * When token renewal is not enabled, this method will not start any periodic tasks. Instead, it
+   * will generate tokens if the driver ref has been provided, update the current user, and send
+   * those tokens to the driver. No future tokens will be generated, so when that initial set
+   * expires, the app will stop working.
+   *
+   * @param driver If provided, the driver where to send the newly generated tokens.
+   *               The same ref will also receive future token updates unless overridden later.
+   * @return The newly logged in user, or null if a principal is not configured.
+   */
+  def start(driver: Option[RpcEndpointRef] = None): UserGroupInformation = {
```

Comment:
Are there going to be unit tests for this?  Because trying to mock lines like `UserGroupInformation.getCurrentUser().getCredentials()` will be tricky. If not, then the way the methods are structured appear to be fine, given successful e2e tests.

Reply:
This code has always lacked proper unit tests. In part it's because it's hard to tell what a good unit test is here, in part because no one ever bothered. It's also lacking integration tests, although I have a few internally that I generally run on these changes.

That needs to be looked at separately; adding changes here without knowing what a good unit test for this code is would be kinda pointless.